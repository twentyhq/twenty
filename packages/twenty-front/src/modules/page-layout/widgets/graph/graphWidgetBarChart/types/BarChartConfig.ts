import { type GraphColorScheme } from '@/page-layout/widgets/graph/types/GraphColorScheme';

export type BarChartConfig = {
  key: string;
  indexValue: string | number;
  gradientId: string;
  colorScheme: GraphColorScheme;
};
