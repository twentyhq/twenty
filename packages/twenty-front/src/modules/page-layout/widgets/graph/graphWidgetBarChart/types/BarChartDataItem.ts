import { type BarDatum } from '@nivo/bar';

export type BarChartDataItem = BarDatum & {
  to?: string;
};
