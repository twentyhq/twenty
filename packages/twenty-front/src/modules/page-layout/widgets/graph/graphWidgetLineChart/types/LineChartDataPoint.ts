export type LineChartDataPoint = {
  x: number | string | Date;
  y: number | null;
  to?: string;
};
