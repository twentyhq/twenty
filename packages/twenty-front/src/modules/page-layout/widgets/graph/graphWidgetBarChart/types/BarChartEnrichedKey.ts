import { type GraphColorScheme } from '@/page-layout/widgets/graph/types/GraphColorScheme';

export type BarChartEnrichedKey = {
  key: string;
  colorScheme: GraphColorScheme;
  label: string;
};
