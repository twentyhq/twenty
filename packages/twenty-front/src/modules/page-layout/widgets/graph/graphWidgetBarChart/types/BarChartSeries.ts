import { type GraphColor } from '@/page-layout/widgets/graph/types/GraphColor';

export type BarChartSeries = {
  key: string;
  label?: string;
  color?: GraphColor;
};
