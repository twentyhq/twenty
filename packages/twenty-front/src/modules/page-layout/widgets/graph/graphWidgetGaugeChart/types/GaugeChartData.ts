import { type GraphColor } from '@/page-layout/widgets/graph/types/GraphColor';

export type GaugeChartData = {
  value: number;
  min: number;
  max: number;
  color?: GraphColor;
  to?: string;
  label?: string;
};
