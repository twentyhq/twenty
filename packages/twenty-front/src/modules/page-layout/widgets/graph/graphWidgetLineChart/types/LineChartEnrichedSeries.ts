import { type LineChartSeries } from '@/page-layout/widgets/graph/graphWidgetLineChart/types/LineChartSeries';
import { type GraphColorScheme } from '@/page-layout/widgets/graph/types/GraphColorScheme';

export type LineChartEnrichedSeries = LineChartSeries & {
  colorScheme: GraphColorScheme;
  gradientId: string;
  shouldEnableArea: boolean;
  label: string;
};
