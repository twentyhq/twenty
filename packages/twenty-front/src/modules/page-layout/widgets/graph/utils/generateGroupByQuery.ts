import gql from 'graphql-tag';

import { type ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { capitalize } from 'twenty-shared/utils';
import { getGroupByQueryName } from '../../../utils/getGroupByQueryName';

export const generateGroupByQuery = ({
  objectMetadataItem,
  aggregateOperations,
}: {
  objectMetadataItem: ObjectMetadataItem;
  aggregateOperations: string[];
}) => {
  const capitalizedSingular = capitalize(objectMetadataItem.nameSingular);
  const queryName = `${capitalize(objectMetadataItem.namePlural)}GroupBy`;
  const queryFieldName = getGroupByQueryName(objectMetadataItem);

  return gql`
    query ${queryName}(
      $groupBy: [${capitalizedSingular}GroupByInput!]!
      $filter: ${capitalizedSingular}FilterInput
      $orderBy: [${capitalizedSingular}OrderByWithGroupByInput!]
      $viewId: UUID
    ) {
      ${queryFieldName}(
        groupBy: $groupBy
        filter: $filter
        orderBy: $orderBy
        viewId: $viewId
      ) {
        groupByDimensionValues${aggregateOperations.length > 0 ? `\n        ${aggregateOperations.join('\n        ')}` : ''}
      }
    }
  `;
};
