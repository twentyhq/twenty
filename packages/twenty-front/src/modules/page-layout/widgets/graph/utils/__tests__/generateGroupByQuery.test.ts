import { type ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { generateGroupByQuery } from '../generateGroupByQuery';

describe('generateGroupByQuery', () => {
  const testCases = [
    {
      description: 'single aggregate operation',
      objectMetadataItem: {
        nameSingular: 'opportunity',
        namePlural: 'opportunities',
      },
      aggregateOperations: ['sumAmountAmountMicros'],
    },
    {
      description: 'multiple aggregate operations',
      objectMetadataItem: {
        nameSingular: 'opportunity',
        namePlural: 'opportunities',
      },
      aggregateOperations: [
        'totalCount',
        'sumAmountAmountMicros',
        'avgAmountAmountMicros',
      ],
    },
    {
      description: 'empty aggregate operations',
      objectMetadataItem: {
        nameSingular: 'person',
        namePlural: 'people',
      },
      aggregateOperations: [],
    },
  ];

  it.each(testCases)(
    'should generate valid GraphQL query for $description',
    ({ objectMetadataItem, aggregateOperations }) => {
      const result = generateGroupByQuery({
        objectMetadataItem: objectMetadataItem as ObjectMetadataItem,
        aggregateOperations,
      });

      expect(result.loc?.source.body).toMatchSnapshot();
    },
  );
});
