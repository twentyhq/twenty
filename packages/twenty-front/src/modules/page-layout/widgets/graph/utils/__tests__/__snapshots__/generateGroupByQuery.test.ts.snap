// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateGroupByQuery should generate valid GraphQL query for empty aggregate operations 1`] = `
"
    query PeopleGroupBy(
      $groupBy: [PersonGroupByInput!]
      $filter: PersonFilterInput
      $orderBy: [PersonOrderByWithGroupByInput!]
      $viewId: UUID
    ) {
      peopleGroupBy(
        groupBy: $groupBy
        filter: $filter
        orderBy: $orderBy
        viewId: $viewId
      ) {
        groupByDimensionValues
      }
    }
  "
`;

exports[`generateGroupByQuery should generate valid GraphQL query for multiple aggregate operations 1`] = `
"
    query OpportunitiesGroupBy(
      $groupBy: [OpportunityGroupByInput!]
      $filter: OpportunityFilterInput
      $orderBy: [OpportunityOrderByWithGroupByInput!]
      $viewId: UUID
    ) {
      opportunitiesGroupBy(
        groupBy: $groupBy
        filter: $filter
        orderBy: $orderBy
        viewId: $viewId
      ) {
        groupByDimensionValues
        totalCount
        sumAmountAmountMicros
        avgAmountAmountMicros
      }
    }
  "
`;

exports[`generateGroupByQuery should generate valid GraphQL query for single aggregate operation 1`] = `
"
    query OpportunitiesGroupBy(
      $groupBy: [OpportunityGroupByInput!]
      $filter: OpportunityFilterInput
      $orderBy: [OpportunityOrderByWithGroupByInput!]
      $viewId: UUID
    ) {
      opportunitiesGroupBy(
        groupBy: $groupBy
        filter: $filter
        orderBy: $orderBy
        viewId: $viewId
      ) {
        groupByDimensionValues
        sumAmountAmountMicros
      }
    }
  "
`;
